_integration-test/run.py

```diff
@@ -0,0 +1,197 @@
+TIMEOUT_SECONDS = 60
+
+
+def poll_for_response(request: str, client: httpx.Client) -> httpx.Response:
+    for i in range(TIMEOUT_SECONDS):
+        response = client.get(
+            request, follow_redirects=True, headers={"Referer": SENTRY_TEST_HOST}
+        )
+        if response.status_code == 200:
+            break
+        time.sleep(1)
+    else:
+        raise AssertionError("timeout waiting for response status code 200")
+    return response
+
+
+@lru_cache
+def get_sentry_dsn(client: httpx.Client) -> str:
+    response = poll_for_response(
+        f"{SENTRY_TEST_HOST}/api/0/projects/sentry/internal/keys/",
+        client,
+    )
+    sentry_dsn = json.loads(response.text)[0]["dsn"]["public"]
+    assert len(sentry_dsn) > 0
+    return sentry_dsn
+
+
+@pytest.fixture(scope="session", autouse=True)
+def configure_self_hosted_environment():
+    response = None
+    subprocess.run(["docker", "compose", "--ansi", "never", "up", "-d"], check=True)
+    for i in range(TIMEOUT_SECONDS):
+        try:
+            response = httpx.get(SENTRY_TEST_HOST, follow_redirects=True)
+        except httpx.ConnectionError:
+            time.sleep(1)
+        if response is not None and response.status_code == 200:
+            break
+    else:
+        raise AssertionError("timeout waiting for self-hosted to come up")
+
+    # Create test user
+    subprocess.run(
+        [
+            "docker",
+            "compose",
+            "exec",
+            "web",
+            "sentry",
+            "createuser",
+            "--force-update",
+            "--superuser",
+            "--email",
+            TEST_USER,
+            "--password",
+            TEST_PASS,
+        ],
+        check=True,
+        text=True,
+        input="y",
+    )
+
+
+@pytest.fixture()
+def client_login():
+    client = httpx.Client()
+    response = client.get(SENTRY_TEST_HOST, follow_redirects=True)
+    parser = BeautifulSoup(response.text, "html.parser")
+    login_csrf_token = parser.find("input", {"name": "csrfmiddlewaretoken"})["value"]
+    login_response = client.post(
+        f"{SENTRY_TEST_HOST}/auth/login/sentry/",
+        follow_redirects=True,
+        data={
+            "op": "login",
+            "username": TEST_USER,
+            "password": TEST_PASS,
+            "csrfmiddlewaretoken": login_csrf_token,
+        },
+        headers={"Referer": f"{SENTRY_TEST_HOST}/auth/login/sentry/"},
+    )
+    assert login_response.status_code == 200
+    yield (client, login_response)
+
+
+def test_initial_redirect():
+    initial_auth_redirect = httpx.get(SENTRY_TEST_HOST, follow_redirects=True)
+    assert initial_auth_redirect.url == f"{SENTRY_TEST_HOST}/auth/login/sentry/"
+
+
+def test_login(client_login):
+    client, login_response = client_login
+    assert '"isAuthenticated":true' in login_response.text
+    assert '"username":"test@example.com"' in login_response.text
+    assert '"isSuperuser":true' in login_response.text
```

Comment:
this is fragile -- this should use some sort of parser to acquire this information (is it json?)

Reply:
it is json inside of an html body, added a html parser and json parser to get this information so it's less fragile